libs/SalesforceSDK/src/com/salesforce/androidsdk/rest/RestClient.java

```diff
@@ -148,35 +157,52 @@ public static void clearOkClientsCache() {
     }
 
 	/**
-	 * Sets the OkHttpclient associated with this user account. The OkHttpclient
+	 * Sets the OkHttpclient.Builder associated with this user account. The OkHttpclient.Builder
 	 * are cached in a map and reused as and when a user account
 	 * switch occurs, to prevent multiple threads being spawned unnecessarily.
 	 */
-	private synchronized void setOkHttpClient() {
-		if (OK_CLIENTS == null) {
-			OK_CLIENTS = new HashMap<>();
+	private synchronized void setOkHttpClientBuilder() {
+		if (OK_CLIENT_BUILDERS == null) {
+			OK_CLIENT_BUILDERS = new HashMap<>();
 		}
 		final String uniqueId = this.oAuthRefreshInterceptor.clientInfo.buildUniqueId();
-		OkHttpClient okHttpClient = null;
+		OkHttpClient.Builder okHttpClientBuilder = null;
 		if (uniqueId != null) {
-			okHttpClient = OK_CLIENTS.get(uniqueId);
-			if (okHttpClient == null) {
-				okHttpClient = httpAccessor.getOkHttpClientBuilder()
-                        .addInterceptor(oAuthRefreshInterceptor)
-                        .build();
-
-                OK_CLIENTS.put(uniqueId, okHttpClient);
+			okHttpClientBuilder = OK_CLIENT_BUILDERS.get(uniqueId);
+			if (okHttpClientBuilder == null) {
+				okHttpClientBuilder = httpAccessor.getOkHttpClientBuilder()
+						.addInterceptor(oAuthRefreshInterceptor);
+				OK_CLIENT_BUILDERS.put(uniqueId, okHttpClientBuilder);
 			}
 		}
-		this.okHttpClient = okHttpClient;
+		this.okHttpClientBuilder = okHttpClientBuilder;
 	}
 
 	/**
-	 * Used by tests
-	 * @param okHttpClient
+	 * Sets the OkHttpclient associated with this user account. The OkHttpclient
+	 * are cached in a map and reused as and when a user account
+	 * switch occurs, to prevent multiple threads being spawned unnecessarily.
 	 */
-	public void setOkHttpClient(OkHttpClient okHttpClient) {
-		this.okHttpClient = okHttpClient;
+	public synchronized void setOkHttpClient(OkHttpClient okHttpClient) {
+		if (OK_CLIENTS == null) {
+			OK_CLIENTS = new HashMap<>();
+		}
+		final String uniqueId = this.oAuthRefreshInterceptor.clientInfo.buildUniqueId();
+		OkHttpClient curOkHttpClient = null;
+		if (uniqueId != null) {
+			curOkHttpClient = OK_CLIENTS.get(uniqueId);
+			if (curOkHttpClient == null) {
```

Comment:
If there's no existing client in the map, check if the passed in one is valid. If not, build a new one. If the passed in one is valid, replace the existing one with the new one. If there is a valid one and the passed in one is invalid, do nothing.